documentation/bolt_types_reference.md

```diff
@@ -0,0 +1,319 @@
+# Bolt data types
+
+## `ApplyResult`
+
+An `ApplyResult` is returned from an [apply action](https://puppet.com/docs/bolt/latest/applying_manifest_blocks.html#)
+as part of a `ResultSet` object and contains information about the apply action.
+
+The following functions are available to `ApplyResult` objects. They are accessed with
+dot notation `$apply_result.function`.
+
+### `action`
+
+**Returns**
+
+`String`: The type of result (`apply`)
+
+### `error`
+
+**Returns**
+
+`Error`: An object constructed from the `_error` field of the result's value.
+
+### `message`
+
+**Returns**
+
+`String`: The `_output` field of the result's value.
+
+### `ok`
+
+**Returns**
+
+`Boolean`: Whether the result was successful.
+
+### `report`
+
+**Returns**
+
+`Hash`: The Puppet report from the apply action.
+
+### `target`
+
+**Returns**
+
+`Target`: The target the result is from.
+
+### `to_data`
+
+**Returns**
+
+`Hash`: A serialized representation of `ApplyResult`.
+
+
+## `Result`
+
+A `Result` is returned from an action executed on a `Target` as part of a `ResultSet`
```

Comment:
```suggestion
A `Result` is returned from an action executed on a single target and contains information about the execution. 
`Result`s are aggregated to make a `ResultSet` object.
```

Reply:
I'm not sure how to say "The return from all your functions is actually going to be ResultSet which is an array of Results" here without being to wordy...but anything else feels confusing.